File: aloelite/pack.py
# ./aloelite/pack.py
# License: Apache-2.0 (disclaimer at bottom of file)
"""
The pack blob codec — a CROSS-IMPLEMENTATION CONTRACT (OP-6/OP-7, TX-2).

A packed subtree is one MsgPack map:

    { "fmt": "aloefs.pack", "ver": 2, "nodes": [ <node>, ... ] }

`nodes` is in TOP-DOWN canonical order (the `subtree` view: depth, then
edge_id, then node_id), one entry per PLACEMENT, parents before children so a
single forward pass restores it. Each node is a map whose keys a writer emits
in exactly this order, omitting the optional ones when absent:

    p   parent's index in `nodes`, or -1 for the root
    t   node type: container | entry | symlink | fifo | socket
    n   the EFFECTIVE name at this placement (coalesce(edge.name, node.name))
    c   created_at, ns
    m   modified_at, ns
    u   uid            (v2; only when set)
    g   gid            (v2; only when set)
    o   mode, 07777    (v2; only when set)
    x   NODE-6 metadata, {string: string}, keys sorted; only when non-empty
    xa  xattrs, {name: bytes}, keys sorted   (v2; only when any)
    rk  retention_keep (v2; leaves only, only when set)
    d   payload bytes; leaves only (a symlink's target, empty for fifo/socket)

Byte rules, so four implementations produce identical blobs: MsgPack with
strings as str and payloads as bin (msgpack-python's `use_bin_type=True`),
integers in their smallest encoding, no absent keys. Both encoders in the tree
follow them by construction; conformance/vectors/pack-v2.json pins them, and
pack-v1.json pins that v1 blobs still read.

What v2 deliberately does NOT carry (doc/DECISIONS.md D-8): atime (noatime
semantics, and get_node coalesces it so a writer cannot tell set from unset),
ctime (the placement trigger owns it), and hardlink identity (each placement
restores as its own node).

VERSIONING. `ver` is a gate, not decoration: a blob written by a newer build
is refused (`unsupported`) rather than read with this build's field set, which
would silently drop whatever the new version added — the same posture db.py
takes toward a newer schema era. A malformed or absent version is `corrupt`.
Older versions stay readable: every field added after v1 is optional on read,
so a v1 blob restores exactly as it did before v2 existed.
"""

from __future__ import annotations

from typing import Any

import msgpack

from .errors import Corrupt, Unsupported

PACK_FMT = "aloefs.pack"
PACK_VER = 2

# The keys a v1 node entry may carry, in emission order.
NODE_KEYS = ("p", "t", "n", "c", "m", "u", "g", "o", "x", "xa", "rk", "d")


def encode(nodes: list[dict[str, Any]]) -> bytes:
    """Serialize `nodes` (already in canonical order, keys in NODE_KEYS
    order) as a pack blob at the current version."""
    return msgpack.packb(
        {"fmt": PACK_FMT, "ver": PACK_VER, "nodes": nodes}, use_bin_type=True
    )


def decode(blob: bytes, **context: object) -> list[dict[str, Any]]:
    """Validate a pack blob and return its node list.

    Raises Corrupt for anything that is not a well-formed pack of a known
    shape, and Unsupported for a pack written by a newer build. `context` is
    attached to the error (the node being unpacked, typically).
    """
    try:
        doc = msgpack.unpackb(blob, raw=False)
    except Exception:
        raise Corrupt("not an aloefs pack blob", **context) from None
    if not isinstance(doc, dict) or doc.get("fmt") != PACK_FMT:
        raise Corrupt("not an aloefs pack blob", **context)
    ver = doc.get("ver")
    if not isinstance(ver, int) or isinstance(ver, bool) or ver < 1:
        raise Corrupt(f"pack blob has no usable version ({ver!r})", **context)
    if ver > PACK_VER:
        raise Unsupported(
            f"pack was written by a newer aloelite (pack format v{ver}; "
            f"this build understands v{PACK_VER}). Upgrade aloelite to "
            f"unpack it.",
            **context,
        )
    nodes = doc.get("nodes")
    if not isinstance(nodes, list):
        raise Corrupt("pack blob has no node list", **context)
    for i, pn in enumerate(nodes):
        if not _well_formed(pn):
            raise Corrupt(f"pack node {i} is malformed", **context)
    return nodes


def _is_int(v: object) -> bool:
    return isinstance(v, int) and not isinstance(v, bool)


def _well_formed(pn: object) -> bool:
    """The node shape: p/t/n required; c/m/u/g/o/rk optional ints; x an
    optional map; xa an optional {str: bytes} map; d optional bytes. Unknown
    keys are ignored (forward-tolerant within a version)."""
    if not isinstance(pn, dict):
        return False
    if not (_is_int(pn.get("p")) and isinstance(pn.get("t"), str)):
        return False
    if not isinstance(pn.get("n"), str):
        return False
    for key in ("c", "m", "u", "g", "o", "rk"):
        if pn.get(key) is not None and not _is_int(pn.get(key)):
            return False
    if pn.get("x") is not None and not isinstance(pn.get("x"), dict):
        return False
    xa = pn.get("xa")
    if xa is not None:
        if not isinstance(xa, dict):
            return False
        if not all(isinstance(k, str) and isinstance(v, bytes) for k, v in xa.items()):
            return False
    return pn.get("d") is None or isinstance(pn.get("d"), bytes)


__all__ = ["PACK_FMT", "PACK_VER", "NODE_KEYS", "encode", "decode"]
# Copyright Michael Godfrey 2026 | aloecraft.org <michael@aloecraft.org>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

File: rust/aloelite-core/src/pack.rs
//! The pack blob codec — a CROSS-IMPLEMENTATION CONTRACT (OP-6/OP-7, TX-2).
//!
//! A packed subtree is one MsgPack map, `{fmt: "aloefs.pack", ver: 2,
//! nodes: [...]}`, with `nodes` in TOP-DOWN canonical order (the `subtree`
//! view: depth, edge_id, node_id), one entry per PLACEMENT, parents before
//! children. Each entry's keys are emitted in [`PackNode`]'s field order,
//! optional ones omitted when absent: `p` parent index or -1, `t` type,
//! `n` the effective placement name, `c`/`m` created/modified ns, then the
//! v2 additions `u`/`g`/`o` (uid, gid, mode) , `x` metadata (keys sorted;
//! only when non-empty), `xa` xattrs (name → bytes, keys sorted), `rk`
//! retention_keep (leaves only), and `d` payload bytes (leaves only).
//! Strings are str, bytes are bin, integers take their smallest encoding —
//! which is what `rmp_serde::to_vec_named` and msgpack-python's
//! `packb(use_bin_type=True)` both do, so the two are byte-identical;
//! `conformance/vectors/pack-v2.json` pins the writer and `pack-v1.json`
//! pins that v1 blobs still read.
//!
//! What v2 deliberately does NOT carry (D-8): atime (noatime semantics, and
//! `get_node` coalesces it so a writer cannot tell set from unset), ctime
//! (the placement trigger owns it), and hardlink identity (each placement
//! restores as its own node).
//!
//! VERSIONING. `ver` is a gate: a blob written by a newer build is
//! `unsupported`, never read with this build's field set (which would drop
//! whatever the new version added, silently). A malformed or absent version
//! is `corrupt`. Older versions stay readable: every field added after v1 is
//! optional on read. The head is decoded before the body so a newer blob's
//! unknown shapes cannot masquerade as corruption.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::errors::{FsError, Result};

// ---------------------------------------------------------------------------
// surface
// ---------------------------------------------------------------------------

pub const PACK_FMT: &str = "aloefs.pack";
pub const PACK_VER: u32 = 2;

/// One node entry. Field order is serialization order and part of the byte
/// contract; `c` and `m` are always written (nil when unknown), everything
/// after them only when present, exactly as the reference writes them.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PackNode {
    pub p: i64,
    pub t: String,
    pub n: String,
    #[serde(default)]
    pub c: Option<i64>,
    #[serde(default)]
    pub m: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub u: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub g: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub o: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub x: Option<BTreeMap<String, String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub xa: Option<BTreeMap<String, Bin>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rk: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub d: Option<Bin>,
}

/// Bytes on the wire: serialized as MsgPack bin, and accepted ONLY from bin.
/// `serde_bytes` would also accept a str or an int sequence, which the
/// reference refuses as `corrupt`; this keeps the two decoders in step.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Bin(pub Vec<u8>);

impl From<Vec<u8>> for Bin {
    fn from(v: Vec<u8>) -> Self {
        Bin(v)
    }
}

impl From<&[u8]> for Bin {
    fn from(v: &[u8]) -> Self {
        Bin(v.to_vec())
    }
}

impl std::ops::Deref for Bin {
    type Target = [u8];
    fn deref(&self) -> &[u8] {
        &self.0
    }
}

/// Serialize `nodes` (already in canonical order) as a v1 pack blob.
pub fn encode(nodes: &[PackNode]) -> Vec<u8> {
    rmp_serde::to_vec_named(&PackDoc {
        fmt: PACK_FMT,
        ver: PACK_VER,
        nodes,
    })
    .expect("encoding to memory cannot fail")
}

/// Validate a pack blob and return its node list: `corrupt` for anything
/// that is not a well-formed pack of a known shape, `unsupported` for a
/// pack written by a newer build.
pub fn decode(blob: &[u8]) -> Result<Vec<PackNode>> {
    // The reference refuses a top-level array; serde would happily read a
    // struct from one positionally, so check the marker first: fixmap,
    // map16 or map32.
    if !matches!(blob.first(), Some(0x80..=0x8f | 0xde | 0xdf)) {
        return Err(FsError::corrupt("not an aloefs pack blob"));
    }
    let head: PackHead =
        rmp_serde::from_slice(blob).map_err(|_| FsError::corrupt("not an aloefs pack blob"))?;
    if head.fmt.as_deref() != Some(PACK_FMT) {
        return Err(FsError::corrupt("not an aloefs pack blob"));
    }
    let ver = match head.ver {
        Some(v) if v >= 1 => v,
        other => {
            return Err(FsError::corrupt(format!(
                "pack blob has no usable version ({other:?})"
            )));
        }
    };
    if ver > i64::from(PACK_VER) {
        return Err(FsError::unsupported(format!(
            "pack was written by a newer aloelite (pack format v{ver}; this build understands v{PACK_VER}). Upgrade aloelite to unpack it."
        )));
    }
    let body: PackBody = rmp_serde::from_slice(blob)
        .map_err(|e| FsError::corrupt(format!("pack blob is malformed: {e}")))?;
    body.nodes
        .ok_or_else(|| FsError::corrupt("pack blob has no node list"))
}

// ---------------------------------------------------------------------------
// depth: the wire shapes
// ---------------------------------------------------------------------------

#[derive(Serialize)]
struct PackDoc<'a> {
    fmt: &'a str,
    ver: u32,
    nodes: &'a [PackNode],
}

/// The version gate's view: read before anything else.
#[derive(Deserialize)]
struct PackHead {
    #[serde(default)]
    fmt: Option<String>,
    #[serde(default)]
    ver: Option<i64>,
}

#[derive(Deserialize)]
struct PackBody {
    #[serde(default)]
    nodes: Option<Vec<PackNode>>,
}

impl Serialize for Bin {
    fn serialize<S: serde::Serializer>(&self, s: S) -> std::result::Result<S::Ok, S::Error> {
        s.serialize_bytes(&self.0)
    }
}

impl<'de> Deserialize<'de> for Bin {
    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> std::result::Result<Self, D::Error> {
        struct V;
        impl serde::de::Visitor<'_> for V {
            type Value = Bin;
            fn expecting(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                f.write_str("msgpack bin")
            }
            fn visit_bytes<E: serde::de::Error>(self, v: &[u8]) -> std::result::Result<Bin, E> {
                Ok(Bin(v.to_vec()))
            }
            fn visit_byte_buf<E: serde::de::Error>(
                self,
                v: Vec<u8>,
            ) -> std::result::Result<Bin, E> {
                Ok(Bin(v))
            }
        }
        d.deserialize_bytes(V)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn node(p: i64, t: &str, n: &str, c: i64, m: i64) -> PackNode {
        PackNode {
            p,
            t: t.into(),
            n: n.into(),
            c: Some(c),
            m: Some(m),
            u: None,
            g: None,
            o: None,
            x: None,
            xa: None,
            rk: None,
            d: None,
        }
    }

    #[test]
    fn bytes_match_the_reference_encoder() {
        // msgpack.packb({"fmt": "aloefs.pack", "ver": 2, "nodes": [
        //   {"p": -1, "t": "container", "n": "d", "c": 5, "m": 5},
        //   {"p": 0, "t": "entry", "n": "f", "c": 6, "m": 7, "o": 420, "x": {"k": "v"},
        //    "xa": {"user.a": b"\x00"}, "rk": 2, "d": b"hi"}]}, use_bin_type=True)
        let leaf = PackNode {
            o: Some(0o644),
            x: Some([("k".to_owned(), "v".to_owned())].into_iter().collect()),
            xa: Some([("user.a".to_owned(), Bin(vec![0]))].into_iter().collect()),
            rk: Some(2),
            d: Some(Bin(b"hi".to_vec())),
            ..node(0, "entry", "f", 6, 7)
        };
        let bytes = encode(&[node(-1, "container", "d", 5, 5), leaf]);
        let expected = concat!(
            "83",
            "a3666d74",
            "ab616c6f6566732e7061636b",
            "a3766572",
            "02",
            "a56e6f646573",
            "92",
            "85",
            "a170",
            "ff",
            "a174",
            "a9636f6e7461696e6572",
            "a16e",
            "a164",
            "a163",
            "05",
            "a16d",
            "05",
            "8a",
            "a170",
            "00",
            "a174",
            "a5656e747279",
            "a16e",
            "a166",
            "a163",
            "06",
            "a16d",
            "07",
            "a16f",
            "cd01a4",
            "a178",
            "81",
            "a16b",
            "a176",
            "a27861",
            "81",
            "a6757365722e61",
            "c40100",
            "a2726b",
            "02",
            "a164",
            "c4026869",
        );
        assert_eq!(hex::encode(&bytes), expected);
        let back = decode(&bytes).unwrap();
        assert_eq!(back.len(), 2);
        assert_eq!(back[1].d.as_deref(), Some(&b"hi"[..]));
        assert_eq!(back[1].o, Some(0o644));
    }

    #[test]
    fn the_gate_runs_before_the_body_and_only_maps_and_bins_are_accepted() {
        // {"fmt": "aloefs.pack", "ver": 3, "nodes": [{"zz": 1}]} -- a v3 node shape we cannot read
        let newer = hex::decode(
            "83a3666d74ab616c6f6566732e7061636ba3766572 03a56e6f64657391 81a27a7a01"
                .replace(' ', ""),
        )
        .unwrap();
        assert_eq!(decode(&newer).err().unwrap().code(), Some("unsupported"));
        // the same node shape at ver 2 is malformed, not newer
        let malformed = hex::decode(
            "83a3666d74ab616c6f6566732e7061636ba3766572 02a56e6f64657391 81a27a7a01"
                .replace(' ', ""),
        )
        .unwrap();
        assert_eq!(decode(&malformed).err().unwrap().code(), Some("corrupt"));
        // {"fmt": "aloefs.pack", "nodes": []} -- no version at all
        let versionless = hex::decode("82a3666d74ab616c6f6566732e7061636ba56e6f64657390").unwrap();
        assert_eq!(decode(&versionless).err().unwrap().code(), Some("corrupt"));
        // ["aloefs.pack", 1, []] -- positionally a valid struct, but not a pack
        let array = hex::decode("93ab616c6f6566732e7061636b0190").unwrap();
        assert_eq!(decode(&array).err().unwrap().code(), Some("corrupt"));
        // a v1 blob still reads
        let v1 = hex::decode("83a3666d74ab616c6f6566732e7061636ba3766572 01a56e6f64657391 85a170ffa174a9636f6e7461696e6572a16ea164a16305a16d05".replace(' ', "")).unwrap();
        assert_eq!(decode(&v1).unwrap()[0].n, "d");
        // a payload given as str, not bin, is refused as the reference refuses it
        let str_payload = hex::decode("83a3666d74ab616c6f6566732e7061636ba3766572 02a56e6f64657391 84a170ffa174a5656e747279a16ea166a164a474657874".replace(' ', "")).unwrap();
        assert_eq!(decode(&str_payload).err().unwrap().code(), Some("corrupt"));
        assert_eq!(
            decode(b"not msgpack").err().unwrap().code(),
            Some("corrupt")
        );
        assert_eq!(decode(b"").err().unwrap().code(), Some("corrupt"));
    }
}
